prov/rxd/src/rxd_cq.c

```diff
@@ -210,11 +210,12 @@ static void rxd_complete_rx(struct rxd_ep *ep, struct rxd_x_entry *rx_entry)
 	struct fi_cq_err_entry err_entry;
 	struct rxd_cq *rx_cq = rxd_ep_rx_cq(ep);
 
-	if ((!(rx_entry->cq_entry.flags & FI_REMOTE_CQ_DATA) &&
-	     (rx_entry->cq_entry.flags & FI_RMA ||
-	     rx_entry->cq_entry.flags & FI_ATOMIC)) ||
-	     rx_entry->flags & (RXD_NO_RX_COMP | RXD_CANCELLED))
+	if (rx_entry->flags & (RXD_NO_RX_COMP | RXD_CANCELLED))
 		goto out;
+	if (!(rx_entry->cq_entry.flags & FI_REMOTE_CQ_DATA) &&
+	     (rx_entry->cq_entry.flags & FI_RMA ||
+	     rx_entry->cq_entry.flags & FI_ATOMIC))
+		goto cntr;
```

Comment:
The use of goto's in this function are driving me nutty.  They simply try to avoid duplicating a single function call.  Please rework to something like this:
```
if (!flags & FI_REMOTE_CQ_DATA) ...) }
    rxd_cntr_report_rx_comp(...);
} else if (bytes_done == len) {
    write_fn(..)
    rxd_cntr_report_rx_comp(...);
} else {
   blah blah
}
rxd_rx_entry_free(...);
```
You can either keep the first if statement (flags & NO_COMP) with a goto

Reply:
Updated